typescript/src/codeActions/custom/renameParameterToNameFromType.ts

```diff
@@ -0,0 +1,96 @@
+
+    return typeSignatureParam.name
+}
+
+const getEdits = (fileName: string, position: number, newText: string, languageService: ts.LanguageService): ts.FileTextChanges[] | undefined => {
+    const renameLocations = languageService.findRenameLocations(fileName, position, false, false, {})
+    if (!renameLocations) return
+
+    const extractFileName = ({ fileName }: ts.RenameLocation) => fileName
+
+    return pipe(
+        groupBy(extractFileName),
+        Object.entries,
+        map(
+            ([fileName, changes]): ts.FileTextChanges => ({
+                fileName,
+                textChanges: changes.map(
+                    ({ textSpan }): ts.TextChange => ({
+                        newText,
+                        span: textSpan,
+                    }),
+                ),
+            }),
+        ),
+    )(renameLocations)
+}
+
+export const renameParameterToNameFromType = {
+    id: 'renameParameterToNameFromType',
+    name: 'Rename Parameter to Name from Type',
+    kind: 'refactor.rewrite.renameParameterToNameFromType',
+    tryToApply(sourceFile, position, range, node, formatOptions, languageService) {
+        if (!node || !position) return
+        const functionSignature = node.parent.parent
+        if (!ts.isIdentifier(node) || !ts.isParameter(node.parent) || !ts.isFunctionLike(functionSignature)) return
+        const { parent: functionDecl, parameters: functionParameters } = functionSignature
+
+        const parameterIndex = functionParameters.indexOf(node.parent)
+        const parameterName = functionParameters[parameterIndex]!.name.getText()
+        const typeParamName = getTypeParamName(functionParameters.indexOf(node.parent), functionDecl, languageService)
+        if (!typeParamName || typeParamName.startsWith('__') || parameterName === typeParamName) return
+
+        if (!formatOptions) return true
+
+        const edits = compact(getEdits(sourceFile.fileName, position, typeParamName, languageService))
+        return {
+            edits,
+        }
+    },
+} satisfies CodeAction
+
+export const renameAllParametersToNameFromType = {
+    id: 'renameAllParametersToNameFromType',
+    name: 'Rename All Parameters to Name from Type',
+    kind: 'refactor.rewrite.renameAllParametersToNameFromType',
+    tryToApply(sourceFile, position, range, node, formatOptions, languageService) {
+        if (!node || !position) return
+        const functionSignature = node.parent.parent
+        if (!ts.isIdentifier(node) || !ts.isParameter(node.parent) || !ts.isFunctionLike(functionSignature)) return
+        const { parent: functionDecl, parameters: functionParameters } = functionSignature
+
+        const paramsToRename = compact(
+            functionParameters.map((functionParameter, index) => {
+                const typeParamName = getTypeParamName(index, functionDecl, languageService)
+                if (!typeParamName) return
+                const isInternal = typeParamName.startsWith('__')
```

Comment:
Not sure if it is the correct way to check the name is internal

Reply:
**YES its okay** this is really uncommon user to name param with two leading underscores. However FYI you could get valueDeclaration of parameter and check wether its identifier. Again, refer to method snippets code https://github.com/zardoy/typescript-vscode-plugins/blob/240f2e1cc89f871d2442d5b28270b38dcd51ab2f/typescript/src/decorateEditsForFileRename.ts